packages/react-components/react-portal/library/src/components/Portal/usePortalMountNode.ts

```diff
@@ -21,71 +20,228 @@ export type UsePortalMountNodeOptions = {
+          return true;
+        }
+
+        return false;
+      },
+    });
+  }, [elementFactory, targetNode, options.disabled]);
+
+  useInsertionEffect!(() => {
+    if (!elementProxy) {
+      return;
+    }
 
-  if (useInsertionEffect) {
-    // eslint-disable-next-line react-hooks/rules-of-hooks
-    useInsertionEffect(() => {
-      if (!element) {
-        return;
-      }
+    const classesToApply = className.split(' ').filter(Boolean);
 
-      const classesToApply = className.split(' ').filter(Boolean);
+    elementProxy.classList.add(...classesToApply);
+    elementProxy.setAttribute('dir', dir);
+    elementProxy.setAttribute('data-portal-node', 'true');
 
-      element.classList.add(...classesToApply);
-      element.setAttribute('dir', dir);
-      element.setAttribute('data-portal-node', 'true');
+    focusVisibleRef.current = elementProxy;
 
-      focusVisibleRef.current = element;
+    return () => {
+      elementProxy.classList.remove(...classesToApply);
+      elementProxy.removeAttribute('dir');
+    };
+  }, [className, dir, elementProxy, focusVisibleRef]);
 
-      return () => {
-        element.classList.remove(...classesToApply);
-        element.removeAttribute('dir');
-      };
-    }, [className, dir, element, focusVisibleRef]);
-  } else {
-    // This useMemo call is intentional for React 17
-    // We don't want to re-create the portal element when its attributes change.
-    // This also should not be done in an effect because, changing the value of css variables
-    // after initial mount can trigger interesting CSS side effects like transitions.
-    // eslint-disable-next-line react-hooks/rules-of-hooks
-    React.useMemo(() => {
-      if (!element) {
-        return;
-      }
+  React.useEffect(() => {
+    return () => {
+      elementProxy?.remove();
+    };
+  }, [elementProxy]);
 
-      // Force replace all classes
-      element.className = className;
-      element.setAttribute('dir', dir);
-      element.setAttribute('data-portal-node', 'true');
+  return elementProxy;
+};
 
-      focusVisibleRef.current = element;
-    }, [className, dir, element, focusVisibleRef]);
+/**
+ * Creates a new element on a "document.body" to mount portals.
+ */
+export const usePortalMountNode = (options: UsePortalMountNodeOptions): HTMLElement | null => {
+  'use no memo';
+
+  const { targetDocument, dir } = useFluent();
+  const mountNode = usePortalMountNodeContext();
+
+  const focusVisibleRef = useFocusVisible<HTMLDivElement>() as React.MutableRefObject<HTMLElement | null>;
+  const classes = usePortalMountNodeStylesStyles();
+  const themeClassName = useThemeClassName();
+
+  const factoryOptions: UseElementFactoryOptions = {
+    dir,
+    disabled: options.disabled,
+    focusVisibleRef,
+
+    className: mergeClasses(themeClassName, classes.root, options.className),
+    targetNode: mountNode ?? targetDocument?.body,
+  };
+
+  if (useInsertionEffect) {
```

Comment:
wondering what consequences this might have for runtime ( conditionally invoking hooks ),

would something like this be more appropriate (the con is that it will create additional `null` in memory for these `useMemo` in both moder/legacy ) ? 

https://github.com/Hotell/fluentui/commit/34581919c87665b12ee58b14d1b1b9ee8e071c4d

Reply:
well that wont mitigate not following rules of hooks.

have you checked my proposed solution ?

Reply:
perfect, this is what I was looking for. ty